vfio/src/vfio_device.rs

```diff
@@ -0,0 +1,594 @@
+        let mut ret =
+            unsafe { ioctl_with_mut_ref(&group, VFIO_GROUP_GET_STATUS(), &mut group_status) };
+        if ret < 0 {
+            return Err(VfioError::GetGroupStatus);
+        }
+
+        if group_status.flags != VFIO_GROUP_FLAGS_VIABLE {
+            return Err(VfioError::GroupViable);
+        }
+
+        let container = VfioContainer::new()?;
+        if container.get_api_version() as u32 != VFIO_API_VERSION {
+            return Err(VfioError::VfioApiVersion);
+        }
+
+        container.check_extension(VFIO_TYPE1v2_IOMMU)?;
+
+        // Safe as we are the owner of group and container_raw_fd which are valid value,
+        // and we verify the ret value
+        let container_raw_fd = container.as_raw_fd();
+        ret = unsafe { ioctl_with_ref(&group, VFIO_GROUP_SET_CONTAINER(), &container_raw_fd) };
+        if ret < 0 {
+            return Err(VfioError::GroupSetContainer);
+        }
+
+        container.set_iommu(VFIO_TYPE1v2_IOMMU)?;
+
+        let device = Self::kvm_device_add_group(vm, &group)?;
+
+        Ok(VfioGroup {
+            group,
+            device,
+            container,
+        })
+    }
+
+    fn kvm_device_add_group(vm: &VmFd, group: &File) -> Result<DeviceFd> {
+        let mut vfio_dev = kvm_bindings::kvm_create_device {
+            type_: kvm_bindings::kvm_device_type_KVM_DEV_TYPE_VFIO,
+            fd: 0,
+            flags: 0,
+        };
+
+        let device_fd = vm
+            .create_device(&mut vfio_dev)
+            .map_err(VfioError::CreateVfioKvmDevice)?;
+
+        let group_fd = group.as_raw_fd();
+        let group_fd_ptr = &group_fd as *const i32;
+        let dev_attr = kvm_bindings::kvm_device_attr {
+            flags: 0,
+            group: kvm_bindings::KVM_DEV_VFIO_GROUP,
+            attr: u64::from(kvm_bindings::KVM_DEV_VFIO_GROUP_ADD),
+            addr: group_fd_ptr as u64,
+        };
+
+        device_fd
+            .set_device_attr(&dev_attr)
+            .map_err(VfioError::KvmSetDeviceAttr)?;
+
+        Ok(device_fd)
+    }
+
+    fn kvm_device_del_group(&self) {
+        let group_fd = self.as_raw_fd();
+        let group_fd_ptr = &group_fd as *const i32;
+        let dev_attr = kvm_bindings::kvm_device_attr {
+            flags: 0,
+            group: kvm_bindings::KVM_DEV_VFIO_GROUP,
+            attr: u64::from(kvm_bindings::KVM_DEV_VFIO_GROUP_DEL),
+            addr: group_fd_ptr as u64,
+        };
+
+        if self.device.set_device_attr(&dev_attr).is_err() {
+            error!("Could not delete VFIO group");
+        }
+    }
+
+    fn unset_container(&self) {
+        let container_raw_fd = self.container.as_raw_fd();
+
+        // Safe as we are the owner of self and container_raw_fd which are valid value.
+        let ret = unsafe { ioctl_with_ref(self, VFIO_GROUP_UNSET_CONTAINER(), &container_raw_fd) };
+        if ret < 0 {
+            error!("Failed to unset container for group");
+        }
```

Comment:
Same question here, are removal problems effectively ignored? Are the reasons for failures from the kernel side effectively never useful?

Reply:
Both `unset_container()` and `device_del_group()` are used from the `Drop` trait implementation. Drop's `drop()` does not return anything, so the propagation would stop here.
However, it does make sense to propagate the last reported kernel error, so I did that. Thanks for the suggestion.